### ISSUE TITLE
[research] Self-improving hypothesis / idea search from u-sekai evidence

### ISSUE THREAD

**Issue by rebuildup** [rebuildup/u-sekai#14]
## Question

How should u-sekai turn evidence from Synthetic User experiments into **better product hypotheses / ideas over repeated iterations**, while avoiding local-optimum collapse, evaluator gaming, benchmark overfitting, and superficial idea diversity?

The goal is not merely to ask an LLM to "suggest improvements from this trace." The research question is whether u-sekai can support a disciplined, evidence-grounded search process in which hypothesis-generation strategies themselves can improve from experiment outcomes.

## Motivation

The current u-sekai vision already contains the larger loop:

```text
observe the world
      ↓
generate ideas / hypotheses
      ↓
build many candidates
      ↓
expose them to users / simulations
      ↓
observe behavior and experience
      ↓
evaluate
      ↓
generate better hypotheses
      ↺
```

u-sekai currently focuses primarily on the interaction / observation / evaluation portion of this loop. However, if implementation becomes cheap, the quality of **which hypotheses are generated next** becomes increasingly important.

A naïve loop such as:

```text
trajectory
→ "suggest an improvement"
→ implement
→ evaluate
→ repeat
```

is likely to converge toward small local edits, overfit known Synthetic User populations, or learn how to satisfy a fixed evaluator rather than discover genuinely better product directions.

Recent work on self-improving agents suggests treating the outer improvement process itself as an optimization problem: preserve multiple candidates, use held-out evaluation, learn from accepted and rejected changes, and force structural rethinking when local improvement stalls.

Reference that prompted this issue:

- https://speakerdeck.com/22mi/how-to-design-self-improvement-for-ai-agents

## Working decomposition

Treat product improvement and idea-generation improvement as two distinct loops.

### 1. Product improvement loop

```text
Product / Environment
        ↓
Synthetic Users
        ↓
trajectories / self-report / telemetry
        ↓
Findings
        ↓
Product hypotheses
        ↓
Candidate changes
        ↓
Product / Environment'
        ↺
```

### 2. Hypothesis-generation improvement loop

```text
Evidence
   ↓
multiple hypothesis-generation strategies
   ↓
hypothesis candidates
   ↓
experiments against generated populations
   ↓
outcomes
   ↓
which hypotheses were useful / predictive / robust?
   ↓
update the search strategy
   ↺
```

The second loop is the focus of this research issue.

### Hypothesis representation

Investigate whether u-sekai should represent ideas as more than free-form text.

A candidate structure may include:

- **Observation** — what actually happened
- **Causal hypothesis** — why it may have happened
- **Intervention** — what should change
- **Prediction** — what should change in observed behavior / experience
- **Target population** — which user conditions the hypothesis concerns
- **Falsifier** — what result should cause rejection or revision
- **Evidence** — supporting trajectories / participant reports / observer findings / telemetry
- **Uncertainty** — what is unknown or weakly supported
- **Lineage** — parent hypotheses / mutations / prior rejected variants

The representation should preserve the distinction between observed fact, interpretation, and proposed intervention.

### Candidate population rather than one "best idea"

Investigate maintaining multiple hypothesis families instead of repeatedly mutating a single winner.

Candidate strategies may deliberately emphasize different evidence:

- behavioral anomalies
- mental-model mismatch
- subjective frustration / distrust
- unexpected workarounds
- accessibility / capability mismatch
- abandonment / hesitation
- high-confidence incorrect beliefs
- disagreement between participant / observer / telemetry
- opportunities rather than failures

The goal is structural diversity, not multiple phrasings of the same suggestion.

### Idea / hypothesis quality dimensions

Avoid immediately defining one universal scalar "idea quality" score.

Candidate dimensions include:

- evidence grounding
- novelty
- structural diversity
- falsifiability
- predicted impact
- population coverage
- robustness across fresh populations
- uncertainty calibration
- prediction accuracy
- unintended side effects
- implementation cost / reversibility (if relevant to the experiment)
- discovery value even when the hypothesis is rejected

Investigate Pareto-style retention where hypotheses that are strong on different dimensions can coexist.

### Prediction calibration

Record:

```text
predicted effect
      vs
observed effect
```

The system should distinguish an idea that sounds plausible from a generator that repeatedly makes well-calibrated, useful predictions.

### Novelty / lineage archive

Investigate storing:

- accepted hypotheses
- rejected hypotheses
- partially supported hypotheses
- failed interventions
- hypothesis lineage
- reason for rejection
- which populations / environments each candidate was tested against

Rejected ideas should remain useful evidence instead of disappearing.

The archive should make it possible to detect when the generator is repeatedly rediscovering semantically equivalent interventions.

### Stagnation / rethink triggers

Investigate explicit triggers for escaping local optimization.

For example, after repeated low novelty / low generalization improvement, the system may temporarily prohibit local edits and require changes at a different abstraction level:

- challenge the causal model
- reframe the user problem
- change the interaction model
- remove the workflow instead of polishing it
- automate the workflow
- change information architecture
- introduce a different entry point
- question whether the feature is needed at all

The trigger should be a property of the search harness rather than a prompt such as "be more creative."

### Proposer / evaluator separation

Investigate separation between:

- hypothesis proposer
- Synthetic User participant
- independent observer
- evaluator
- experiment / population generator

A proposer should not automatically see every hidden evaluation population or evaluator detail.

### Anti-overfitting / anti-Goodhart design

Investigate:

- training / visible populations
- held-out Synthetic User populations
- fresh generated populations
- cross-model evaluation
- blind or partially blind baseline-vs-candidate comparison
- periodic calibration against real-user evidence where available

A self-improving loop that learns to satisfy one fixed judge is a failure mode, even if its reported metric improves.

### Meta-optimization target

Clarify what u-sekai may optimize over time.

Possible mutable artifacts include:

- hypothesis-generation prompts / Skills
- analysis strategies
- evidence-selection strategies
- critique strategies
- candidate branching rules
- diversity constraints
- stagnation rules
- evaluation protocols

The product implementation agent itself does **not** need to live inside u-sekai.

## Initial non-scope

- Building an autonomous product-development system now
- Letting u-sekai merge or deploy product changes automatically
- Choosing a final optimization algorithm
- Choosing a final LLM / model provider
- Defining one universal UX score
- Treating Synthetic User results as ground truth
- Replacing real-user research
- Committing to GEPA, evolutionary search, Bayesian optimization, RL, or another specific optimizer before comparison
- Implementing this before the core Synthetic User evidence model is sufficiently understood

## Relationship to existing u-sekai research

This topic depends on or should consume results from:

- subjective participant / observer / telemetry evidence design
- open-ended user-state / story / population generation
- Synthetic-vs-real calibration
- generative benchmark / anti-overfitting design
- mental models, forgetting, and evolving cognitive state

It should also feed back into the broader vision's `generate better hypotheses ↺` loop.

Because research numbering is currently being reorganized in the core-vision work and `R-10` is already used by the AgentENV issue on `main`, this issue intentionally does **not** claim an R-number yet.

## Suggested investigation

1. Survey self-improving agent / prompt / Skill optimization work for reusable mechanisms, including:
   - candidate populations
   - reflective mutation
   - held-out validation
   - accepted / rejected edit history
   - Pareto retention
   - novelty / diversity preservation
   - stagnation detection
2. Define a minimal hypothesis schema that separates evidence, causal interpretation, intervention, prediction, and falsifier.
3. Create example u-sekai traces and test whether multiple generators produce structurally different hypotheses.
4. Design a method for detecting semantic duplicates / local-edit collapse.
5. Define an experiment protocol that tests a hypothesis against visible and held-out generated populations.
6. Measure prediction calibration in addition to post-hoc evaluator preference.
7. Explore how failed / rejected ideas should influence subsequent search.
8. Compare:
   - single-best iterative refinement
   - multi-candidate search
   - Pareto / archive-based search
   - forced reframe after stagnation
9. Identify which parts belong in u-sekai core versus replaceable external strategy modules.
10. Document failure modes where self-improvement gives a misleading appearance of progress.

## Acceptance criteria

- A clear boundary exists between **product improvement** and **hypothesis-generation improvement**.
- A candidate hypothesis representation is documented.
- At least two multi-candidate search strategies are compared against single-best iterative refinement.
- Structural novelty / duplicate detection is investigated.
- Held-out / fresh-population evaluation is defined so the proposer cannot trivially optimize directly against all evaluation cases.
- Proposer / participant / observer / evaluator responsibilities are separated.
- Stagnation / rethink criteria are proposed.
- Accepted, rejected, and partially supported hypotheses have a retention / lineage model.
- Prediction calibration is treated as a measurable property.
- Anti-Goodhart / benchmark-overfitting risks and mitigations are documented.
- The boundary between u-sekai and external coding / design / deployment agents is explicit.
- Any implementation follow-ups are opened only after this research establishes a justified architecture.
